src/compiler/checker.ts

```diff
@@ -7771,16 +7771,34 @@ namespace ts {
                 if (target.flags & TypeFlags.Union && containsType(targetTypes, source)) {
                     return Ternary.True;
                 }
-                const len = targetTypes.length;
-                for (let i = 0; i < len; i++) {
-                    const related = isRelatedTo(source, targetTypes[i], reportErrors && i === len - 1);
+                for (const type of targetTypes) {
+                    const related = isRelatedTo(source, type, /*reportErrors*/ false);
                     if (related) {
                         return related;
                     }
                 }
+                const discriminantType = findMatchingDiscriminantType(source, target);
+                isRelatedTo(source, discriminantType || targetTypes[targetTypes.length - 1], reportErrors);

```

Comment:
Enclose the two lines above in `if (reportErrors) { ... }` and pass `true` as the `reportErrors` argument.